### PR TITLE
handle invalid accept-charset in requests - default to utf-8

### PR DESCRIPTION
**Pull request by pjfanning**
see #300 

I've added #585 to look into the XML case - this PR focuses on text/plain.

### CODE REVIEW

File: http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala [hunk 1]
```diff
@@ -275,4 +275,23 @@ class MarshallingDirectivesSpec extends RoutingSpec with Inside {
...
+    "render text with requested encoding if an `Accept-Charset` request header requests a non-UTF-8 encoding" in {
+      Get() ~> `Accept-Charset`(`ISO-8859-1`) ~> complete(foo) ~> check {
+        responseEntity shouldEqual HttpEntity(ContentType(`text/plain`, `ISO-8859-1`), foo)
+      }
+    }
+    "render text with UTF-8 encoding if an `Accept-Charset` request header requests an unknown encoding" in {
+      Get() ~> `Accept-Charset`(HttpCharset("unknown")(Nil)) ~> complete(foo) ~> check {
+        responseEntity shouldEqual HttpEntity(ContentType(`text/plain`, `UTF-8`), foo)
+      }
+    }
```
Comment (raboof) [inline http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:295]: I realize this is a comment on the current behavior, and not on your change, but this seems strange to me: it looks like the default string marshaller will return the string passed to it as `text/plain` claiming it is in whatever encoding that was requested by the user, without doing anything to make sure the string is actually in that encoding.

This is only correct if you assume the implementation correctly took into account the charset, which seems extremely unlikely. I wonder if it wouldn't make more sense to leave out the optional `charset` parameter to the return content type in this case. If people have specific requirements (which seems somewhat unlikely) they can specify the charset explicitly.

This would be a behavior change, but it seems like a reasonable one if we note it in the release notes.

Comment (pjfanning) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:295]: before the change to the main source code in this PR, this test actually fails - the response is an error response

Comment (michaelf-crwd) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: Hi, 

Thanks for having a look at this bug / feature request. I recently sent in a duplicate issue (https://github.com/apache/pekko-http/issues/583) unaware of this one. I worked on a similar PR the past two days, and have now found and looked through your changes - that would seem to fix it nicely! Great. 

For your consideration, I'd like to suggest adding in an extra unit-test in "MarshallingSpac.scala" with something along the lines of:

```
val invalidAcceptCharsetHeader = `Accept-Charset`(HttpCharsetRange(HttpCharset.custom("invalid")))
val request = HttpRequest().withHeaders(Seq(invalidAcceptCharsetHeader))
marshalToResponse("Hello", request).entity.contentType.charsetOption should be(Some(HttpCharsets.`UTF-8`))
```

Some refactoring thoughts; Perhaps the try-catch in "PredefinedToEntityMarshallers.scala" could be more subtly placed in "HttpCharset.scala" as:

```
/** Returns this HttpCharset instance if the provided nio charset is available or a desired default otherwise */
  def withNioCharsetOrElse(default: HttpCharset): HttpCharset = {
    if (_nioCharset.isSuccess) {
      // Return this instance, as the provided nioCharset did not result in a java.nio.charset.UnsupportedCharsetException
      this
    } else {
      default
    }
  }
```

and then have the "PredefinedToEntityMarshallers.scala" do:

```
Marshaller.withOpenCharset(mediaType) { (s, cs) => {
  HttpEntity(mediaType.withCharset(cs.withNioCharsetOrElse(HttpCharsets.`UTF-8`)), s)
} }
```

Looking forward to seeing your changes merged.

Thanks again and best regards,
Michael

Comment (pjfanning) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: Thanks @michaelf-crwd - I have added a new HttpCharsets function similar to what you suggested (c8514e78919d4bc41cc459aee9cb88a0a94043a6).
I will look later at adding extra tests like the one you suggested.

Comment (pjfanning) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: @raboof would you be able to take a look at this again? I don't think this change breaks anything and makes the pekko-http support for `Accept-Charset` a bit more tolerant.

Comment (pjfanning) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: @mdedetrich would you have time to look at this? It's basically the last item to decide on before a 1.1.0 release.

Comment (mdedetrich) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: Ill check this tomorrow, there is context I need to grock

**Review by nvollmar — APPROVED** [review]
lgtm

**Comment by raboof** [comment]
Sorry for not finding time to look into this sooner, will do so now.

File: http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala (hunk 1 above)
Comment (jrudolph) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: > without doing anything to make sure the string is actually in that charset

@raboof, WDYM with that? Java strings are basically UTF-16 encoded, so they are known to be unicode. What else could go wrong?

**Review by jrudolph — APPROVED** [review]
lgtm, I agree with the idea to treat the `Accept-XYZ` headers rather softly, especially `Accept-Charset` which is losing relevancy these days. As long as the server clearly indicates what representation it uses, it seems fine.

**Review by raboof — APPROVED** [review]
I tested various scenario's:

# Explicit charset in response

i.e. a route with `complete(HttpEntity(ContentTypes.`text/html(UTF-8)`, "Say Hällö to Pekko HTTP"))`:

- request: `curl -v localhost:8080/hello`
  old behavior: response in Content-Type: text/html; charset=UTF-8 :green_heart: 
  new behavior: same :green_heart: 
- request: `curl -v -H "Accept-Charset: blop" localhost:8080/hello`
  old behavior: 500 error and stacktrace in the logs :stop_sign: 
  new behavior: 406 Not Acceptable, "Resource representation is only available with these types: text/html; charset=UTF-8" :green_heart: 
- request: `curl -v -H "Accept-Charset: US-ASCII" localhost:8080/hello`
  old behavior: 406 Not Acceptable, "Resource representation is only available with these types: text/html; charset=UTF-8" :green_heart: 
  new behavior: same :green_heart: 
- curl -v -H "Accept-Charset: ISO-8859-1" localhost:8080/hello
  old -> response 406 Not Acceptable :green_heart: 
  new -> same :green_heart: 

# No explicit charset in response

i.e. a route with `complete("Say Hällö to Pekko HTTP")`:

with `complete("Say Hällö to Pekko HTTP")`
- request: `curl -v localhost:8080/hello`
  old behavior: response in `text/plain; charset=UTF-8` :green_heart: 
  new behavior: same :green_heart: 
- request: `curl -v -H "Accept-Charset: blop" localhost:8080/hello`
  old behavior: 500 error and stacktrace in the logs :stop_sign: 
  new behavior: response in `text/plain; charset=UTF-8` :green_heart: 
- request: `curl -v -H "Accept-Charset: US-ASCII" localhost:8080/hello`
  old behavior: response with `Content-Type: text/plain; charset=US-ASCII` header, ~but UTF-8 body~ response converted to US-ASCII ~:stop_sign:~ :green_heart: 
  new behavior: same ~:stop_sign:~ :green_heart: 
- curl -v -H "Accept-Charset: ISO-8859-1" localhost:8080/hello
  old -> response with `Content-Type: text/plain; charset=ISO-8859` header, ~but UTF-8 body~ response converted to US-ASCII ~:stop_sign:~ :green_heart: 
  new -> same ~:stop_sign:~ :green_heart: 

In summary, I think this PR is a definite improvement. ~Some behavior seems wrong to me, but that behavior was *already* like that, so that doesn't need to delay merging this PR or releasing 1.1.0. I'll put my code where my mouth is and propose a fix for the behavior that looks wrong to me, but that's independent of 1.1.0. Sorry for taking so long before confirming ;)~ (actually the strings are converted taking into the encodings just fine, sorry about the noise :facepalm: )

File: http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala (hunk 1 above)
Comment (raboof) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: what I meant is, when the route does `complete(some_unicode_string)`, and the marshalling logic results in a `Content-Type: text/plain; charset=ISO-8859-1` header, that response seems wrong: the header promises the body will be in ISO-8859-1, but the body is actually in unicode. I think it would make more sense to make sure we produce a `Content-Type: text/plain; charset=UTF-8` header unless the developer explicitly marked the response as being in a different charset.

Comment (raboof) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: or will the marshalling logic actually convert the unicode to the charset from the content type? in that case of course it's all good.

Comment (raboof) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: Indeed it does, sorry about the noise :facepalm:

Comment (jrudolph) [reply http-tests/src/test/scala/org/apache/pekko/http/scaladsl/server/directives/MarshallingDirectivesSpec.scala:300]: :+1:

**Comment by jrudolph** [comment]
> the strings are converted to bytes taking into account the encodings just fine, all good, sorry about the noise 🤦

The whole negotiation story is really sophisticated and tries to take that into account, indeed. It's also one of the performance hotspots of the routing layer.

**Comment by jrudolph** [comment]
Anyway thanks for that comprehensive testing session, @raboof :+1:

**Comment by pjfanning** [comment]
Thanks everyone. I'll merge as is but if anyone has any requests for changes, I'll look at them.
